### ISSUE TITLE
boxfuse config file variable substitution (changed?)

### ISSUE THREAD

**Issue by jtheuer** [boxfuse/boxfuse-issues#112]
What are the rules for variable substitution in ```boxfuse.conf``` files?

I noticed in one of my ec2 console logs that ```$BOXFUSE_ENV``` isn't always substituted (worked only in the second case):

```
NEW_RELIC_APP_NAME=foo-$BOXFUSE_ENV
SPRING_PROFILES_ACTIVE=production
```

Config file:
```
envvars.NEW_RELIC_APP_NAME=foo-$BOXFUSE_ENV
envvars.SPRING_PROFILES_ACTIVE=$BOXFUSE_ENV
```

As far as I remember: It worked in another project where I used a ```boxfuse.conf``` file and an older version of the boxfuse client version.

**Comment by axelfontaine** [comment]
You are correct and we aren't fully happy with the current behavior.

The (undocumented) rules currently are that if the first character is a $ then substitution will be performed, otherwise not. This was done to avoid triggering substitution in the middle of passwords containing $ for example.

One alternative could be to force the users to escape any special characters. This is a little confusing in the beginning though as failures (such as in the password case) may be rather hard to track down

Another option (which we currently favor) would be to explicitly specify that substitution is desired. Our current idea would be to do this by adding a $ at the end of the variable name (think of it as $= operator instead of =).

Example:

```
envvars.NEW_RELIC_APP_NAME$=foo-$BOXFUSE_ENV
```

would define an environment variable named `NEW_RELIC_APP_NAME` (no trailing $) where the value would be `foo-bar` when $BOXFUSE_ENV=bar

and

```
envvars.NEW_RELIC_APP_NAME=foo-$BOXFUSE_ENV
```

would define an environment variable named `NEW_RELIC_APP_NAME` where the value would always literally be `foo-$BOXFUSE_ENV`

What do you think? Do you see better alternatives or issues with this proposal?

**Comment by jtheuer** [comment]
I see the issue, there is probably no perfect way. The $ at the end would work but I've never seen that pattern before.

I think, always interpreting $VAR and either force users to escape the ```$``` or using ``` ' ... ' ``` quotes would be the most familiar pattern for me, like ```envvars.FOO='my $pa55w0rd '```

Jan

**Comment by jtheuer** [comment]
Would also allow lines like ```envvars.FOO='my password has spaces    '  # and this is a comment```. It should probably also support ```"``` style quotes with variable substitution, then it works like in bash.

**Comment by axelfontaine** [comment]
I am afraid this might open a can of worms as the escaping complexity in bash is just atrocious. See http://stackoverflow.com/questions/1250079/how-to-escape-single-quotes-within-single-quoted-strings

**Comment by jtheuer** [comment]
Good point. The only configuration file I know that natively supports variable substitution is maven. They replace ```${foo}``` in their pom.xml if it is defined, otherwise the plain text will be used. I can't say I like that behavior, better fail-fast.

Since I only use it with ```$BOXFUSE_ENV``` I created a dedicated ```boxfuse-production.conf``` file and overwrote that property there.
